cai_causal_graph/causal_graph.py

```diff
@@ -534,6 +542,75 @@ def is_empty(self) -> bool:
         """Return True if there are no nodes and edges. False otherwise."""
         return len(self.nodes) == 0 and len(self.edges) == 0
 
+    def _check_node(self, identifier: NodeLike) -> str:
+        identifier = Node.identifier_from(identifier)
+
+        if identifier in self._nodes_by_identifier:
+            raise CausalGraphErrors.NodeDuplicatedError(f'Node already exists: {identifier}')
+        return identifier
+
+    def _prepare_nodes(self, source: NodeLike, destination: NodeLike) -> Tuple[List[Node], List[Node]]:
+        """Prepare the source and destination nodes for adding an edge."""
```

Comment:
Again, would be good to describe what happens in this preparation. Also describe what the return is

Reply:
I agree. It's not necessary for protected methods, but maybe it's worth here to add parameter and return docstrings?